<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Farhad Shokoohi" />

<meta name="date" content="2017-07-26" />

<title>DMCHMM: Differentially Methylated CpG using Hidden Markov Model</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">DMCHMM: Differentially Methylated CpG using Hidden Markov Model</h1>
<h4 class="author"><em>Farhad Shokoohi</em></h4>
<h4 class="date"><em>2017-07-26</em></h4>



<p>DNA methylation studies have increased in number over the past decade thanks to the recent advances in next-generation sequencing (NGS) and microarray technology (MA), providing many data sets at high resolution, enabling researchers to understand methylation patterns and their regulatory roles in biological processes and diseases.<br />
Notwithstanding that diverse methods and software have created ample opportunities for researchers to do quantitative analysis, they make it difficult for practitioners to choose the one that is suitable and efficient in analyzing DNA methylation data. Having examined most of differentially methylation identification tools for bisulfite sequencing (BS-Seq) data, we observed several drawbacks in the existing analytic tools. To address these issues we have developed a novel differentially methylated CpG site identification tool which is based on Hidden Markov models (HMM) called <code>DMCHMM</code>. This vignette provides some guidelines on how to use the package and analyze BS-Seq data.</p>
<p>Following topics will be discussed in this vignette:</p>
<ul>
<li><strong>Reading BS-Seq data</strong></li>
<li><strong>Simulating BS-Seq data</strong></li>
<li><strong>Predicting methylation levels using HMM and EM algorithm</strong></li>
<li><strong>Predicting methylation levels using HMM and MCMC algorithm</strong></li>
<li><strong>Identifying DMCs</strong></li>
</ul>
<div id="s4-classes" class="section level2">
<h2>S4 Classes</h2>
<p>Two different classes are defined by extending the <code>SummarizedExperiment-class</code>. The <code>BSData-class</code> is designed to hold BS-Seq data. Similarly, <code>BSData-method</code> is defined to create a <code>BSData</code> object. This class includes two slots: the <code>methReads</code>, a matrix with columns representing samples and rows representing genomic positions (CpG sites) and elements of matrix representing methylation counts at each position in each sample; the <code>totalReads</code>, a matrix with similar columns and rows except the elements representing total number of reads.</p>
</div>
<div id="reading-bs-seq-data" class="section level2">
<h2>Reading BS-Seq data</h2>
<p>For reading raw BS-Seq data we adopted The <code>readBismark</code> function from <code>BiSeq</code> package. The <code>readBismark-method</code> reads samples stored in different files with six columns of <em>chromosome</em>, <em>start position</em>, <em>end position</em>, <em>methylation percentage</em>, <em>number of Cs</em> and <em>number of Ts</em>.</p>
<p>Three data files are included in the <code>DMCHMM</code> package for illustration. The data can be imported using following code.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(DMCHMM)
fn &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;DMCHMM&quot;</span>))
fn.f &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="dt">package=</span><span class="st">&quot;DMCHMM&quot;</span>), <span class="dt">full.names=</span><span class="ot">TRUE</span>)
OBJ &lt;-<span class="st"> </span><span class="kw">readBismark</span>(fn.f, fn)
cdOBJ &lt;-<span class="st"> </span><span class="kw">DataFrame</span>(<span class="dt">Cell =</span> <span class="kw">factor</span>(<span class="kw">c</span>(<span class="st">&quot;BC&quot;</span>, <span class="st">&quot;TC&quot;</span>,<span class="st">&quot;Mono&quot;</span>),
<span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;BC&quot;</span>, <span class="st">&quot;TC&quot;</span>, <span class="st">&quot;Mono&quot;</span>)), <span class="dt">row.names =</span> <span class="kw">c</span>(<span class="st">&quot;BCU1568&quot;</span>,<span class="st">&quot;BCU173&quot;</span>,<span class="st">&quot;BCU551&quot;</span>))
<span class="kw">colData</span>(OBJ) &lt;-<span class="st"> </span>cdOBJ
OBJ</code></pre></div>
<pre><code> output: 
    class: BSData 
    dim: 25668 3 
    metadata(0):
    assays(2): totalReads methReads
    rownames(25668): 1 2 ... 25667 25668
    rowData names(0):
    colnames(3): BCU1568 BCU173 BCU551
    colData names(1): Cell</code></pre>
</div>
<div id="simulating-bs-seq-data" class="section level2">
<h2>Simulating BS-Seq data</h2>
<p>The above data set only include one sample for each cell type. We need more samples to be able to compare their methylations and find DMCs. For illustration we generate a sample of BS-Seq data as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nr &lt;-<span class="st"> </span><span class="dv">500</span>; nc &lt;-<span class="st"> </span><span class="dv">8</span>
metht &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">as.integer</span>(<span class="kw">runif</span>(nr *<span class="st"> </span>nc, <span class="dv">0</span>, <span class="dv">100</span>)), nr)
methc &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rbinom</span>(<span class="dt">n=</span>nr*nc,<span class="kw">c</span>(metht),<span class="dt">prob =</span> <span class="kw">runif</span>(nr*nc)),nr,nc)
r1 &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="kw">rep</span>(<span class="st">&quot;chr1&quot;</span>, nr), <span class="kw">IRanges</span>(<span class="dv">1</span>:nr, <span class="dt">width=</span><span class="dv">1</span>), <span class="dt">strand=</span><span class="st">&quot;*&quot;</span>)
<span class="kw">names</span>(r1) &lt;-<span class="st"> </span><span class="dv">1</span>:nr
cd1 &lt;-<span class="st"> </span><span class="kw">DataFrame</span>(<span class="dt">Group=</span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;G1&quot;</span>,<span class="st">&quot;G2&quot;</span>),<span class="dt">each=</span>nc/<span class="dv">2</span>),<span class="dt">row.names=</span>LETTERS[<span class="dv">1</span>:nc])
OBJ1 &lt;-<span class="st"> </span><span class="kw">BSData</span>(<span class="dt">rowRanges=</span>r1,<span class="dt">methReads=</span>methc,<span class="dt">totalReads=</span>metht,<span class="dt">colData=</span>cd1)
OBJ1</code></pre></div>
<pre><code>output: 
    class: BSData 
    dim: 500 8 
    metadata(0):
    assays(4): methReads totalReads
    rownames(500): 1 2 ... 499 500
    rowData names(0):
    colnames(8): A B ... G H
    colData names(1): Group</code></pre>
</div>
<div id="predicting-methylation-levels-using-hmm-and-em-algorithm" class="section level2">
<h2>Predicting methylation levels using HMM and EM algorithm</h2>
<p>There are two approaches to smoothed the data before testing for DMCs. Either EM or MCMC can be used to predict methylation levels utilizing HMM. The <code>methHMEM-method</code> which is developed to predict methylation levels. The output is a <code>BSDMCs-class</code> that can be either used to find DMCs or use MCMC algorithm to re-smooth the raw data. The process is as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">OBJ2 &lt;-<span class="st"> </span><span class="kw">methHMEM</span>(OBJ1, <span class="dt">mc.cores=</span><span class="kw">detectCores</span>())
OBJ2</code></pre></div>
<pre><code>output: 
    class: BSDMCs 
    dim: 500 8 
    metadata(3): K Beta Pm
    assays(4): methReads totalReads methLevels methStates
    rownames(500): 1 2 ... 499 500
    rowData names(0):
    colnames(8): A B ... G H
    colData names(1): Group</code></pre>
</div>
<div id="predicting-methylation-levels-using-hmm-and-mcmc-algorithm" class="section level2">
<h2>Predicting methylation levels using HMM and MCMC algorithm</h2>
<p>Although EM algorithm is a fast way to smooth the data but the results are not as good as the MCMC algorithm. The MCMC algorithm, however, is slow. In order to increase the speed, we first use <code>methHMEM-method</code> to find the HMM order for each sample and then we use <code>methHMCMC-method</code> to predict methylation levels. The procedure is as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">OBJ3 &lt;-<span class="st"> </span><span class="kw">methHMMCMC</span>(OBJ2, <span class="dt">mc.cores=</span><span class="kw">detectCores</span>())
OBJ3</code></pre></div>
<pre><code>output: 
    class: BSDMCs 
    dim: 500 8 
    metadata(3): K Beta Pm
    assays(4): methReads totalReads methLevels methStates
    rownames(500): 1 2 ... 499 500
    rowData names(0):
    colnames(8): A B ... G H
    colData names(1): Group</code></pre>
</div>
<div id="identifying-dmcs" class="section level2">
<h2>Identifying DMCs</h2>
<p>Having smoothed the data using HMM, we run linear between predicted methylation levels and grouping covariate. In case other covariates exist, one can use the <code>formula</code> argument to specify a linear model. When there is no covariates no action is required. The following command identifys the DMCs. The results are stored in a <code>BSDMCs-class</code> and can be retrived by calling <code>metadata</code> command.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">OBJ4 &lt;-<span class="st"> </span><span class="kw">findDMCs</span>(OBJ3, <span class="dt">mc.cores=</span><span class="kw">detectCores</span>())
<span class="kw">head</span>(<span class="kw">metadata</span>(OBJ4)$DMCHMM)</code></pre></div>
<pre><code>output: 
      DMCs   pvalues   qvalues DMCsG1vsG2 methDirG1vsG2
    1    0 0.9666323 0.9947863          0         equal
    2    0 0.8364271 0.9939796          0         hyper
    3    0 0.2407768 0.9793929          0         hyper
    4    0 0.6684606 0.9924782          0          hypo
    5    0 0.3632399 0.9862448          0         equal
    6    0 0.4307814 0.9883764          0          hypo</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
